Update 2018-10-23-react-v-16-6.md (#179)
reactjs/fr.reactjs.org: content/blog/2018-10-23-react-v-16-6.md

---
title: "React v16.6.0: lazy, memo et contextType"
author: [sebmarkbage]
---

> Nous publions aujourd'hui React 16.6 avec quelques nouvelles fonctionnalités pratiques. Une forme de PureComponent / shouldComponentUpdate pour les fonctions composants, un moyen de diviser le code à l'aide de Suspense et un moyen plus simple d'utiliser Context à partir des composants à base de classes.

Reviewer: ```suggestion
Nous publions aujourd'hui React 16.6 avec quelques nouvelles fonctionnalités pratiques. Une forme de `PureComponent` / `shouldComponentUpdate` pour les fonctions composants, un moyen de découper le code à l'aide de Suspense et un moyen plus simple d'utiliser `Context` à partir des composants à base de classes.
```

> Découvrez le [changelog](#changelog) complet en bas.

Reviewer: ```suggestion
Découvrez le [changelog](#changelog) complet plus bas.
```

## [`React.memo`](/docs/react-api.html#reactmemo) {#reactmemo}

> Les composants à base de classes peuvent éviter le rendu lorsque leurs entrées en props sont les mêmes en utilisant [`PureComponent`](/docs/react-api.html#reactpurecomponent) ou [`shouldComponentUpdate`](/docs/react-component.html#shouldcomponentupdate). Maintenant, vous pouvez faire la même chose avec les fonctions composants en les encapsulant dans [`React.memo`](/docs/react-api.html#reactmemo).

Reviewer: ```suggestion
Les composants à base de classes peuvent éviter le rendu lorsque les valeurs de leurs props sont les mêmes en utilisant [`PureComponent`](/docs/react-api.html#reactpurecomponent) ou [`shouldComponentUpdate`](/docs/react-component.html#shouldcomponentupdate). Désormais, vous pouvez faire la même chose avec les fonctions composants en les enrobant avec [`React.memo`](/docs/react-api.html#reactmemo).
```

```js
const MyComponent = React.memo(function MyComponent(props) {

>   /* le rendu se fait lorsque props change */

Reviewer: ```suggestion
  /* le rendu ne se fait que lorsque les props changent */
```

});
```

> ## [`React.lazy`](/docs/code-splitting.html#reactlazy): La division de code avec `Suspense` {#reactlazy-code-splitting-with-suspense}

Reviewer: ```suggestion
## [`React.lazy`](/docs/code-splitting.html#reactlazy) : la découpe de code avec `Suspense` {#reactlazy-code-splitting-with-suspense}
```

> Vous avez peut-être vu [le discours de Dan sur React Suspense à JSConf Iceland](/blog/2018/03/01/sneak-peek-beyond-react-16.html). Vous pouvez maintenant utiliser le composant Suspense pour faire la [division de code](/docs/code-splitting.html#reactlazy) en encapsulant une importation dynamique dans un appel à `React.lazy()`.

Reviewer: ```suggestion
Vous avez peut-être vu [la présentation de Dan sur React Suspense à JSConf Iceland](/blog/2018/03/01/sneak-peek-beyond-react-16.html). Vous pouvez maintenant utiliser le composant Suspense pour faire de la [découpe de code](/docs/code-splitting.html#reactlazy) en enrobant une importation dynamique avec un appel à `React.lazy()`.
```

```js

> import React, {lazy, Suspense} from 'react';

Reviewer: Dans le code ci-dessous, remplace `Loading...` par `Chargement…` (véritable caractère d'ellipse au passage, copie-colle) : on traduit les contenus affichés.

}
```

> Le composant Suspense permettra également aux auteurs de bibliothèques de commencer à construire la récupération de données avec le support de Suspense à l'avenir.

Reviewer: ```suggestion
Le composant Suspense permettra également aux auteur·e·s de bibliothèques de commencer à construire de la récupération de données prenant en charge Suspense à l'avenir.
```

> > Note: Cette fonctionnalité n'est pas encore disponible pour le rendu côté serveur. Le support de Suspense sera ajouté dans une version ultérieure.

Reviewer: ```suggestion
> Note : cette fonctionnalité n'est pas encore disponible pour le rendu côté serveur. La prise en charge de Suspense sera ajoutée dans une version ultérieure.
```

## [`static contextType`](/docs/context.html#classcontexttype) {#static-contexttype}

> Dans [React 16.3](/blog/2018/03/29/react-v-16-3.html) nous avons introduit Context API en remplacement de la précédente [Legacy Context](/docs/legacy-context.html) API.

Reviewer: ```suggestion
Dans [React 16.3](/blog/2018/03/29/react-v-16-3.html) nous avons introduit l’API de Contexte en remplacement de la précédente API [Legacy Context](/docs/legacy-context.html).
```

```js
const MyContext = React.createContext();
```

> Nous avons reçu des commentaires selon lesquels l’adoption de la nouvelle API d’application de rendu peut être difficile dans les composants à base de classes. Nous avons donc ajouté une API de commodité à [consommer une valeur de contexte depuis un composant à base de classe](/docs/context.html#classcontexttype).

Reviewer: ```suggestion
Nous avons reçu des commentaires selon lesquels l’adoption de la nouvelle API à base de props de rendu peut s’avérer difficile dans les composants à base de classes. Nous avons donc ajouté une API de confort pour [consommer une valeur de contexte depuis un composant à base de classe](/docs/context.html#classcontexttype).
```

```js
class MyClass extends React.Component {
  static contextType = MyContext;
  componentDidMount() {
    let value = this.context;

>     /* effectuer un effet secondaire lors du montage en utilisant la valeur de MyContext */

Reviewer: ```suggestion
    /* réaliser un effet secondaire lors du montage en utilisant la valeur de MyContext */
```

  }
  render() {
    let value = this.context;

>     /* rendre quelque chose basé sur la valeur de MyContext */

Reviewer: ```suggestion
    /* afficher quelque chose basé sur la valeur de MyContext */
```

  }
}
```

## [`static getDerivedStateFromError()`](/docs/react-component.html#static-getderivedstatefromerror) {#static-getderivedstatefromerror}

> React 16 introduit les [Périmètres d’erreurs](/blog/2017/07/26/error-handling-in-react-16.html) pour gérer les erreurs générées dans les rendus React. Nous avions déjà la méthode de cycle de vie `componentDidCatch` qui est déclenchée après une erreur. C'est formidable pour la journalisation des erreurs sur le serveur. Il vous permet également d'afficher une interface utilisateur différente pour l'utilisateur en appelant `setState`.

Reviewer: ```suggestion
React 16 a introduit les [Périmètres d’erreurs](/blog/2017/07/26/error-handling-in-react-16.html) pour gérer les erreurs lancées au sein des rendus React. Nous avions déjà la méthode de cycle de vie `componentDidCatch` qui est déclenchée après une erreur. C'est formidable pour la journalisation des erreurs sur le serveur. Ça vous permet également d'afficher une interface utilisateur différente pour l'utilisateur en appelant `setState`.
```

> Avant que cela ne soit déclenché, nous rendons `null` à la place de l'arbre qui a généré une erreur. Cela casse parfois les composants parents qui ne s’attendent pas à ce que leurs références soient vides. La récupération des erreurs sur le serveur ne fonctionne pas non plus, car les méthodes du cycle de vie `Did` ne se déclenchent pas lors du rendu côté serveur.

Reviewer: ```suggestion
Avant de la déclencher, nous faisons le rendu de `null` à la place de l'arbre qui a généré une erreur. Ça casse parfois les composants parents qui ne s’attendent pas à ce que leurs refs soient vides. La récupération des erreurs sur le serveur ne fonctionne pas non plus, puisque les méthodes du cycle de vie `Did` ne sont pas déclenchées lors du rendu côté serveur.
```

> Nous ajoutons une autre méthode d'erreur qui vous permet de rendre l'interface utilisateur de secours avant la fin du rendu. Voir la documentation pour [`getDerivedStateFromError()`](/docs/react-component.html#static-getderivedstatefromerror).

Reviewer: ```suggestion
Nous ajoutons une autre méthode d'erreur qui vous permet de produire une interface utilisateur de secours avant la fin du rendu. Consultez la documentation de [`getDerivedStateFromError()`](/docs/react-component.html#static-getderivedstatefromerror).
```

> > Note: `getDerivedStateFromError()` n'est pas encore disponible pour le rendu côté serveur. Il est conçu pour fonctionner avec le rendu côté serveur dans une version ultérieure. Nous le publions tôt pour que vous puissiez commencer à vous préparer à l'utiliser.

Reviewer: ```suggestion
> Note : `getDerivedStateFromError()` n'est pas encore disponible pour le rendu côté serveur. Elle est conçue pour fonctionner avec le rendu côté serveur dans une version ultérieure. Nous la publions tôt pour que vous puissiez commencer à vous préparer à l'utiliser.
```

> ## Déprécations en StrictMode {#deprecations-in-strictmode}

Reviewer: ```suggestion
## Dépréciations en StrictMode {#deprecations-in-strictmode}
```

> Dans [16.3](/blog/2018/03/29/react-v-16-3.html#strictmode-component) nous avons introduit le composant [`StrictMode`](/docs/strict-mode.html). Il vous permet d'activer les alertes précoces pour les modèles susceptibles de poser problème à l'avenir.

Reviewer: ```suggestion
La version [16.3](/blog/2018/03/29/react-v-16-3.html#strictmode-component) a introduit le composant [`StrictMode`](/docs/strict-mode.html). Il vous permet d'activer des alertes précoces pour les approches susceptibles de poser problème à l'avenir.
```

> Nous avons ajouté deux API supplémentaires à la liste des API obsolètes de `StrictMode`. Si vous n'utilisez pas `StrictMode` vous n'avez pas à vous inquiéter; ces avertissements ne tireront pas pour vous.

Reviewer: ```suggestion
Nous avons ajouté deux API supplémentaires à la liste des API obsolètes en `StrictMode`. Si vous n'utilisez pas `StrictMode` vous n'avez pas à vous inquiéter ; ces avertissements ne se déclencheront pas pour vous.
```

> * __ReactDOM.findDOMNode()__ - Cette API est souvent mal comprise et la plupart des utilisations de celle-ci sont inutiles. Il peut aussi être étonnamment lent dans React 16. [Voir les docs](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) pour les chemins de mise à niveau possibles.

Reviewer: ```suggestion
* __ReactDOM.findDOMNode()__ - Cette API est souvent mal comprise et la plupart des utilisations de celle-ci sont inutiles. Elle peut aussi être étonnamment lente dans React 16. [Consultez les docs](/docs/strict-mode.html#warning-about-deprecated-finddomnode-usage) pour les mécanismes de mise à niveau possibles.
```

> * __Legacy Context__ en utilisant contextTypes et getChildContext - Legacy context rend React légèrement plus lent et plus gros que nécessaire. C’est pourquoi nous souhaitons vivement encourager la mise à niveau vers le [nouveau context API](/docs/context.html). Espérons que l’ajout de [`contextType`](/docs/context.html#classcontexttype) API rend cela un peu plus facile.

Reviewer: ```suggestion
* __Legacy Context__ en utilisant contextTypes et getChildContext - Le Legacy Context rend React légèrement plus lent et plus gros que nécessaire. C’est pourquoi nous souhaitons vivement encourager la mise à niveau vers la [nouvelle API de Contexte](/docs/context.html). Espérons que l’ajout de l’API [`contextType`](/docs/context.html#classcontexttype) rendra ça un peu plus facile.
```

Si vous rencontrez des difficultés pour effectuer la mise à niveau, nous aimerions connaître votre avis.

## Installation {#installation}

> React v16.6.0 est disponible dans le registre npm.

Reviewer: ```suggestion
React v16.6.0 est disponible sur npm.
```

> Pour installer React 16 avec Yarn, exécutez:

Reviewer: ```suggestion
Pour installer React 16 avec Yarn, exécutez :
```

```bash
yarn add react@^16.6.0 react-dom@^16.6.0
```

> Pour installer React 16 avec npm, exécutez:

Reviewer: ```suggestion
Pour installer React 16 avec npm, exécutez :
```

```bash
npm install --save react@^16.6.0 react-dom@^16.6.0
```

Nous fournissons également des versions UMD de React via un CDN:

```html
<script crossorigin src="https://unpkg.com/react@16/umd/react.production.min.js"></script>
<script crossorigin src="https://unpkg.com/react-dom@16/umd/react-dom.production.min.js"></script>
```

Reportez-vous à la documentation pour les [instructions d'installation détaillées](/docs/installation.html).

## Changelog {#changelog}

### React {#react}

* Ajouter `React.memo()` comme alternative à `PureComponent` pour les fonctions. ([@acdlite](https://github.com/acdlite) dans [#13748](https://github.com/facebook/react/pull/13748))

> * Ajouter `React.lazy()` pour les composants de division de code. ([@acdlite](https://github.com/acdlite) dans [#13885](https://github.com/facebook/react/pull/13885))

Reviewer: ```suggestion
* Ajouter `React.lazy()` pour la découpe de code les composants. ([@acdlite](https://github.com/acdlite) dans [#13885](https://github.com/facebook/react/pull/13885))
```

> * `React.StrictMode` averti à propos de legacy context API ([@bvaughn](https://github.com/bvaughn) dans [#13760](https://github.com/facebook/react/pull/13760))

Reviewer: ```suggestion
* `React.StrictMode` avertit désormais du recours à l’API Legacy Context ([@bvaughn](https://github.com/bvaughn) dans [#13760](https://github.com/facebook/react/pull/13760))
```

> * `React.StrictMode` averti à propos de `findDOMNode`. ([@sebmarkbage](https://github.com/sebmarkbage) dans [#13841](https://github.com/facebook/react/pull/13841))

Reviewer: ```suggestion
* `React.StrictMode` avertit désormais du recours à `findDOMNode`. ([@sebmarkbage](https://github.com/sebmarkbage) dans [#13841](https://github.com/facebook/react/pull/13841))
```

> * Renommer `unstable_AsyncMode` à `unstable_ConcurrentMode`. ([@trueadm](https://github.com/trueadm) dans [#13732](https://github.com/facebook/react/pull/13732))

Reviewer: ```suggestion
* Renommer `unstable_AsyncMode` en `unstable_ConcurrentMode`. ([@trueadm](https://github.com/trueadm) dans [#13732](https://github.com/facebook/react/pull/13732))
```

> * Renommer `unstable_Placeholder` à `Suspense`, et `delayMs` à `maxDuration`. ([@gaearon](https://github.com/gaearon) dans [#13799](https://github.com/facebook/react/pull/13799) et [@sebmarkbage](https://github.com/sebmarkbage) dans [#13922](https://github.com/facebook/react/pull/13922))

Reviewer: ```suggestion
* Renommer `unstable_Placeholder` en `Suspense`, et `delayMs` en `maxDuration`. ([@gaearon](https://github.com/gaearon) dans [#13799](https://github.com/facebook/react/pull/13799) et [@sebmarkbage](https://github.com/sebmarkbage) dans [#13922](https://github.com/facebook/react/pull/13922))
```

### React DOM {#react-dom}

> * Ajouter `contextType` comme moyen plus ergonomique de s’abonner au contexte d’une classe ([@bvaughn](https://github.com/bvaughn) dans [#13728](https://github.com/facebook/react/pull/13728))

Reviewer: ```suggestion
* Ajouter `contextType` comme moyen plus ergonomique de s’abonner au Contexte depuis une classe ([@bvaughn](https://github.com/bvaughn) dans [#13728](https://github.com/facebook/react/pull/13728))
```

> * Ajouter `getDerivedStateFromError` méthode de cycle de vie permettant de détecter les erreurs dans un futur rendu asynchrone côté serveur. ([@bvaughn](https://github.com/bvaughn) dans [#13746](https://github.com/facebook/react/pull/13746))

Reviewer: ```suggestion
* Ajouter la méthode de cycle de vie `getDerivedStateFromError` pour détecter les erreurs à l’avenir lors du rendu asynchrone côté serveur. ([@bvaughn](https://github.com/bvaughn) dans [#13746](https://github.com/facebook/react/pull/13746))
```

* Avertir quand `<Context>` est utilisé à la place de `<Context.Consumer>`. ([@trueadm](https://github.com/trueadm) dans [#13829](https://github.com/facebook/react/pull/13829))
* Correction du calque gris sur iOS Safari. ([@philipp-spiess](https://github.com/philipp-spiess) dans [#13778](https://github.com/facebook/react/pull/13778))

> * Corrige un bug causé par un écrasement `window.event` en développement. ([@sergei-startsev](https://github.com/sergei-startsev) dans [#13697](https://github.com/facebook/react/pull/13697))

Reviewer: ```suggestion
* Corrige un bug causé par un écrasement de `window.event` en développement. ([@sergei-startsev](https://github.com/sergei-startsev) dans [#13697](https://github.com/facebook/react/pull/13697))
```

### React DOM Server {#react-dom-server}

> * Ajouter un support pour `React.memo()`. ([@alexmckenley](https://github.com/alexmckenley) dans [#13855](https://github.com/facebook/react/pull/13855))

Reviewer: ```suggestion
* Ajouter la prise en charge de `React.memo()`. ([@alexmckenley](https://github.com/alexmckenley) dans [#13855](https://github.com/facebook/react/pull/13855))
```

> * Ajouter un support pour `contextType`. ([@alexmckenley](https://github.com/alexmckenley) et [@sebmarkbage](https://github.com/sebmarkbage) dans [#13889](https://github.com/facebook/react/pull/13889))

Reviewer: ```suggestion
* Ajouter la prise en charge de `contextType`. ([@alexmckenley](https://github.com/alexmckenley) et [@sebmarkbage](https://github.com/sebmarkbage) dans [#13889](https://github.com/facebook/react/pull/13889))
```

> ### Scheduler (Experimental) {#scheduler-experimental}

Reviewer: ```suggestion
### Planificateur (expérimental) {#scheduler-experimental}
```

> * Renommez le paquet en `scheduler`. ([@gaearon](https://github.com/gaearon) dans [#13683](https://github.com/facebook/react/pull/13683))

Reviewer: ```suggestion
* Renommer le module en `scheduler`. ([@gaearon](https://github.com/gaearon) dans [#13683](https://github.com/facebook/react/pull/13683))
```

* Prise en charge des niveaux de priorité, des continuations et des rappels encapsulés. ([@acdlite](https://github.com/acdlite) dans [#13720](https://github.com/facebook/react/pull/13720) et [#13842](https://github.com/facebook/react/pull/13842))
* Améliorer le mécanisme de secours dans les environnements non-DOM. ([@acdlite](https://github.com/acdlite) dans [#13740](https://github.com/facebook/react/pull/13740))

> * Planifie `requestAnimationFrame` antérieurement. ([@acdlite](https://github.com/acdlite) dans [#13785](https://github.com/facebook/react/pull/13785))

Reviewer: ```suggestion
* Planifier `requestAnimationFrame` plus tôt. ([@acdlite](https://github.com/acdlite) dans [#13785](https://github.com/facebook/react/pull/13785))
```

> * Corrigez la détection du DOM pour qu'elle soit plus complète. ([@trueadm](https://github.com/trueadm) dans [#13731](https://github.com/facebook/react/pull/13731))

Reviewer: ```suggestion
* Correction de la détection du DOM pour qu'elle soit plus complète. ([@trueadm](https://github.com/trueadm) dans [#13731](https://github.com/facebook/react/pull/13731))
```

* Correction de bugs avec le traçage d'interaction. ([@bvaughn](https://github.com/bvaughn) dans [#13590](https://github.com/facebook/react/pull/13590))

> * Ajouter le `envify` transform au paquet. ([@mridgway](https://github.com/mridgway) dans [#13766](https://github.com/facebook/react/pull/13766))

Reviewer: ```suggestion
* Ajouter la transformée `envify` au module. ([@mridgway](https://github.com/mridgway) dans [#13766](https://github.com/facebook/react/pull/13766))
```
